From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <kfian294ma4@gmail.com>
Date: Fri, 8 Nov 2024 19:35:49 +0000
Subject: [PATCH] Sakura: Optimise check inside blocks and traverse blocks

Original license: MIT
Original project: https://github.com/Samsuik/Sakura

Chunk cache is already removed from this patch, since we already have it in Leaf

The JMH benchmark of this patch can be found in SunBox's `BetterBlocksTraverse`

diff --git a/net/minecraft/world/entity/monster/Ghast.java b/net/minecraft/world/entity/monster/Ghast.java
index 673d4e06d6dc4c94ddaafcdc86cbdbd686964ddd..667e51bdd9bf54bec9bcf5fa9a80b16ddbd99131 100644
--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -332,7 +332,7 @@ public class Ghast extends Mob implements Enemy {
             AABB boundingBox = this.ghast.getBoundingBox();
             AABB aabb = boundingBox.move(delta);
             if (this.careful) {
-                for (BlockPos blockPos : BlockPos.betweenClosed(aabb.inflate(1.0))) {
+                for (BlockPos blockPos : org.dreeam.leaf.util.list.SimpleBlockPosIterator.iterable(aabb.inflate(1.0))) { // Leaf - Sakura - optimise check inside blocks
                     if (!this.blockTraversalPossible(this.ghast.level(), null, null, blockPos, false, false)) {
                         return false;
                     }
diff --git a/net/minecraft/world/entity/projectile/ThrowableProjectile.java b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
index dfda0560e364aa7777bc1aa76febc55f1c2dd5a0..f55c2d40f62048b2d62ab3539929b05a6a2f4ad6 100644
--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -93,7 +93,7 @@ public abstract class ThrowableProjectile extends Projectile {
 
     private void handleFirstTickBubbleColumn() {
         if (this.firstTick) {
-            for (BlockPos blockPos : BlockPos.betweenClosed(this.getBoundingBox())) {
+            for (BlockPos blockPos : org.dreeam.leaf.util.list.SimpleBlockPosIterator.iterable(this.getBoundingBox())) { // Leaf - Sakura - optimise check inside blocks
                 BlockState blockState = this.level().getBlockState(blockPos);
                 if (blockState.is(Blocks.BUBBLE_COLUMN)) {
                     blockState.entityInside(this.level(), blockPos, this, InsideBlockEffectApplier.NOOP, true);
diff --git a/net/minecraft/world/level/BlockGetter.java b/net/minecraft/world/level/BlockGetter.java
index 8a21910fe7ed61c903c4f36fced63b179ace4577..f59b2201b1ac260d9076bf73ca1a337dd782afb4 100644
--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -212,7 +212,7 @@ public interface BlockGetter extends LevelHeightAccessor {
     static boolean forEachBlockIntersectedBetween(Vec3 from, Vec3 to, AABB boundingBox, BlockGetter.BlockStepVisitor visitor) {
         Vec3 vec3 = to.subtract(from);
         if (vec3.lengthSqr() < Mth.square(1.0E-5F)) {
-            for (BlockPos blockPos : BlockPos.betweenClosed(boundingBox)) {
+            for (BlockPos blockPos : org.dreeam.leaf.util.list.SimpleBlockPosIterator.iterable(boundingBox)) { // Leaf - Sakura - optimise check inside blocks
                 if (!visitor.visit(blockPos, 0)) {
                     return false;
                 }
@@ -220,6 +220,20 @@ public interface BlockGetter extends LevelHeightAccessor {
 
             return true;
         } else {
+            // Leaf start - Sakura - optimise check inside blocks
+            final boolean xZero = vec3.x() == 0.0;
+            final boolean yZero = vec3.y() == 0.0;
+            final boolean zZero = vec3.z() == 0.0;
+            if (xZero && yZero || yZero && zZero || xZero && zZero) {
+                int blockIndex = 0;
+                for (final BlockPos blockPos : org.dreeam.leaf.util.list.SimpleBlockPosIterator.traverseBoundsInDirection(vec3, boundingBox, 16.0)) {
+                    if (!visitor.visit(blockPos, blockIndex++)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            // Leaf end - Sakura - optimise check inside blocks
             LongSet set = new LongOpenHashSet();
 
             for (BlockPos blockPos1 : BlockPos.betweenCornersInDirection(boundingBox.move(vec3.scale(-1.0)), vec3)) {
@@ -234,7 +248,7 @@ public interface BlockGetter extends LevelHeightAccessor {
             if (i < 0) {
                 return false;
             } else {
-                for (BlockPos blockPos2 : BlockPos.betweenCornersInDirection(boundingBox, vec3)) {
+                for (BlockPos blockPos2 : org.dreeam.leaf.util.list.SimpleBlockPosIterator.iterable(boundingBox)) { // Leaf - Sakura - optimise check inside blocks
                     if (set.add(blockPos2.asLong()) && !visitor.visit(blockPos2, i + 1)) {
                         return false;
                     }
